keras/backend/common/dtypes.py

```diff
@@ -0,0 +1,310 @@
+import functools
+
+from keras import backend
+from keras.api_export import keras_export
+from keras.backend.common.variables import ALLOWED_DTYPES
+
+# Default dtypes corresponding to Python scalars.
+PYTHON_SCALAR_DTYPES = {
+    bool: "bool",
+    int: "int64",
+    float: "float64",
+    # TODO: support complex dtypes
+    # complex: "complex128",
+}
+BIT64_TO_BIT16_DTYPE = {
+    "int32": "int16",
+    "int64": "int16",
+    "uint32": "uint16",
+    "uint64": "uint16",
+    "float32": "float16",
+    "float64": "float16",
+    # TODO: support complex dtypes
+    # "complex64": "complex32",
+    # "complex128": "complex32",
+}
+BIT64_TO_BIT32_DTYPE = {
+    "int64": "int32",
+    "uint64": "uint32",
+    "float64": "float32",
+    # TODO: support complex dtypes
+    # "complex128": "complex64",
+}
+
+
+@functools.lru_cache(maxsize=None)
+def canonicalize_dtype(float_type, dtype):
+    if float_type == "float16":
+        return BIT64_TO_BIT16_DTYPE.get(dtype, dtype)
+    elif float_type == "float32":
+        return BIT64_TO_BIT32_DTYPE.get(dtype, dtype)
+    elif float_type == "float64":
+        return dtype
+    else:
+        raise ValueError(
+            f"Invalid float_type: {float_type}, `float_type` must be one of"
+            f"('float16', 'float32', 'float64')."
+        )
+
+
+def dtype(x, *, canonicalize=False):
+    """Return the dtype object for a tensor or type."""
+    dtype = None
+    if x is None:
+        raise ValueError(f"Invalid argument to dtype: {x}.")
+    # when inputing python scalar or tensor
+    elif type(x) in PYTHON_SCALAR_DTYPES:
+        dtype = PYTHON_SCALAR_DTYPES[type(x)]
+    elif backend.is_tensor(x):
+        dtype = backend.standardize_dtype(x.dtype)
+    # when inputing python dtype or tensor dtype
+    elif isinstance(x, type) and x in PYTHON_SCALAR_DTYPES:
+        dtype = PYTHON_SCALAR_DTYPES[x]
+    elif hasattr(x, "name"):
+        dtype = x.name
+    elif hasattr(x, "__str__") and "torch" in str(dtype):
+        dtype = str(x).split(".")[-1]
+    # when inputing dtype string
+    elif x in ALLOWED_DTYPES:
+        dtype = x
+    elif x in ["int", "float"]:
+        if x == "int":
+            dtype = "int64"
```

Comment:
Should we default int to int64, or to int32? JAX doesn't support int64 by default so it may be better to default to int32.

Reply:
I have changed to utilize standardize_dtype